### DISCUSSION TITLE
Async Transaction still active after cancellation

### DISCUSSION THREAD

**Discussion by mikec1z** [psycopg/psycopg#555] [General]
Postgres Version 15.2

```
>>> psycopg.__version__
'3.1.8'
```

```
try:
        asyncio.current_task().cancel()
        await cursor.execute("SELECT 1")
catch asyncio.CancelledError as e:
        print(connection.info.transaction_status)
 
```

The output of the above code is
> TransactionStatus.ACTIVE

Even after explicitly cancelling the transaction with `connection.pgconn.get_cancel().cancel()` it still remains in  the `ACTIVE`  state. It seems whenever any query is interrupted with a cancellation it remains in ACTIVE.

I can't determine if this is a bug in postres, psycopg, or if I don't correctly how to cancel an active transaction.

**Comment by dvarrazzo** [comment]
It is a bug, already reported in #543. It's already fixed in master, but not released yet.

I will try to make a release this week. If you can check out master for a test, that would be great.

**Comment by mikec1z** [comment]
Verified for new build  

```
    import psycopg
    print(psycopg.__version__)
    try:
        asyncio.current_task().cancel()
        await cursor.execute("SELECT 1")
    except asyncio.CancelledError as e:
        print(connection.info.transaction_status)
```

> 3.1.9
> TransactionStatus.INTRANS
>

**Comment by mikec1z** [comment]
I'm still coming across a few cases where the transaction remains in ACTIVE when execute() is interrupted with task cancellation. I'm trying to reduce one instance to a minimal code example, but it seems to require cancellation at a very precise timing. I'll post here when I have it.

**Reply by dvarrazzo** [reply]
Thank you, that's appreciated.

**Comment by mikec1z** [comment]
By changing the psycopg code and injecting a cancellation directly within the execute() path, I was able to create an example that consistently reproduces the problem. You may or may not consider this valid, but if nothing else it's an example where we show the transaction is ACTIVE after execute() is interrupted.

1) Edit waiting.py insert `asyncio.current_task().cancel()` before `ev.wait()` in wait_async

```
async def wait_async(gen: PQGen[RV], fileno: int) -> RV:
....
                asyncio.current_task().cancel()
                await ev.wait()
```

2) Run this code

```
    try:
        await cursor.execute("SELECT 1")
    except asyncio.CancelledError as e:
        pass # expected
    finally:
        print(connection.info.transaction_status)
```

Output

> TransactionStatus.ACTIVE

**Reply by dvarrazzo** [reply]
Are you editing `wait_async()` in order to create a race condition?

I don't say it's not valid, but, have you got a suggestion about avoiding the problem?

By the look of it, it seems that you execute `task.cancel()` before the query is flushed to the server, so we issue `pgconn.get_cancel()` before the server has a query on the way, and the cancel turns out to be no-op. Then, at the first movement on the socket, the query reaches the server.

Playing a bit more with the moment in which the cancel is received using this script:

<details>
<summary>Test script `test-555.py`</summary>

```python
import asyncio
import psycopg
import psycopg.waiting
from psycopg import pq

wait_async_orig = psycopg.waiting.wait_async


async def wait_async_hacked(gen, fileno):
    errors_at = 0

    from psycopg.waiting import Event, get_event_loop, WAIT_R, WAIT_W, READY_R, READY_W

    ev = Event()
    loop = get_event_loop()

    def wakeup(state) -> None:
        nonlocal ready
        ready |= state
        ev.set()

    try:
        if errors_at <= 0:
            asyncio.current_task().cancel()
        errors_at -= 1

        s = next(gen)
        while True:
            reader = s & WAIT_R
            writer = s & WAIT_W
            if not reader and not writer:
                raise e.InternalError(f"bad poll status: {s}")
            ev.clear()
            ready = 0
            if reader:
                loop.add_reader(fileno, wakeup, READY_R)
            if writer:
                loop.add_writer(fileno, wakeup, READY_W)
            try:
                await ev.wait()
            finally:
                if reader:
                    loop.remove_reader(fileno)
                if writer:
                    loop.remove_writer(fileno)

            if errors_at <= 0:
                asyncio.current_task().cancel()
            errors_at -= 1
            s = gen.send(ready)

    except StopIteration as ex:
        rv = ex.args[0] if ex.args else None
        return rv


async def main():
    conn = await psycopg.AsyncConnection.connect("host=localhost")
    conn.pgconn.trace(2)
    conn.pgconn.set_trace_flags(pq.Trace.SUPPRESS_TIMESTAMPS | pq.Trace.REGRESS_MODE)

    psycopg.waiting.wait_async = wait_async_hacked
    try:
        cur = await conn.execute("select pg_sleep(0.1)")
        print(await cur.fetchone())
    except asyncio.CancelledError:
        print("cancelled")

    print(f"{conn.info.transaction_status=}")


try:
    asyncio.run(main())
except asyncio.CancelledError:
    print("main cancelled")
```

</details>

With different values of `error_at`:

<details>
<summary>0 (like your example): the cancel is sent before the query. It is the condition you verify:</summary>

```
$ PSYCOPG_IMPL=python python test-555.py 
F	13	Parse	 "" "BEGIN" 0
F	14	Bind	 "" "" 0 0 1 0
F	6	Describe	 P ""
F	9	Execute	 "" 0
F	4	Sync
B	4	ParseComplete
B	4	BindComplete
B	4	NoData
B	10	CommandComplete	 "BEGIN"
B	5	ReadyForQuery	 T
F	25	Query	 "select pg_sleep(0.1)"
cancelled
conn.info.transaction_status=<TransactionStatus.ACTIVE: 1>
F	4	Terminate
```
</details>

<details>
<summary>1: cancel is managed as expected. Further code in psycopg handles the cancel error response</summary>

```
$ PSYCOPG_IMPL=python python test-555.py 
F	13	Parse	 "" "BEGIN" 0
F	14	Bind	 "" "" 0 0 1 0
F	6	Describe	 P ""
F	9	Execute	 "" 0
F	4	Sync
B	4	ParseComplete
B	4	BindComplete
B	4	NoData
B	10	CommandComplete	 "BEGIN"
B	5	ReadyForQuery	 T
F	25	Query	 "select pg_sleep(0.1)"
B	33	RowDescription	 1 "pg_sleep" NNNN 0 NNNN 4 -1 0
B	NN	ErrorResponse	 S "ERROR" V "ERROR" C "57014" M "canceling statement due to user request" F "SSSS" L "SSSS" R "SSSS" \x00
B	5	ReadyForQuery	 E
cancelled
conn.info.transaction_status=<TransactionStatus.INERROR: 3>
F	4	Terminate
main cancelled
```

</details>

<details>
<summary>2: cancelling happens after the query has successfully finished the roundtrip, so the state is consistent again</summary>

```
$ PSYCOPG_IMPL=python python test-555.py 
F	13	Parse	 "" "BEGIN" 0
F	14	Bind	 "" "" 0 0 1 0
F	6	Describe	 P ""
F	9	Execute	 "" 0
F	4	Sync
B	4	ParseComplete
B	4	BindComplete
B	4	NoData
B	10	CommandComplete	 "BEGIN"
B	5	ReadyForQuery	 T
F	25	Query	 "select pg_sleep(0.1)"
B	33	RowDescription	 1 "pg_sleep" NNNN 0 NNNN 4 -1 0
B	10	DataRow	 1 0 ''
B	13	CommandComplete	 "SELECT 1"
B	5	ReadyForQuery	 T
('',)
conn.info.transaction_status=<TransactionStatus.INTRANS: 2>
F	4	Terminate
main cancelled
```

</details>

I don't know if there is a way to avoid a cancel to hit between when we have loaded the query in the libpq buffer and when we flush it...

**Reply by dvarrazzo** [reply]
Peppering the `AsyncConnection.wait()` function with debug info:

```diff
$ git diff psycopg/psycopg/connection_async.py
diff --git a/psycopg/psycopg/connection_async.py b/psycopg/psycopg/connection_async.py
index 24904807..1a9a2895 100644
--- a/psycopg/psycopg/connection_async.py
+++ b/psycopg/psycopg/connection_async.py
@@ -345,15 +345,22 @@ class AsyncConnection(BaseConnection[Row]):
 
     async def wait(self, gen: PQGen[RV]) -> RV:
         try:
+            print("wait 1", self.info.transaction_status)
             return await waiting.wait_async(gen, self.pgconn.socket)
+            print("wait 2")
         except (asyncio.CancelledError, KeyboardInterrupt):
+            print("wait 3", self.info.transaction_status)
             # On Ctrl-C, try to cancel the query in the server, otherwise
             # the connection will remain stuck in ACTIVE state.
             c = self.pgconn.get_cancel()
             c.cancel()
+            print("wait 4", self.info.transaction_status)
             try:
+                print("wait 5", self.info.transaction_status)
                 await waiting.wait_async(gen, self.pgconn.socket)
+                print("wait 6")
             except e.QueryCanceled:
+                print("wait 7")
                 pass  # as expected
             raise
```

the log of the case 0 above show that the connection is in ACTIVE state and doesn't change from there:

```
$ PSYCOPG_IMPL=python python test-555.py 
wait 1 TransactionStatus.IDLE
F	13	Parse	 "" "BEGIN" 0
F	14	Bind	 "" "" 0 0 1 0
F	6	Describe	 P ""
F	9	Execute	 "" 0
F	4	Sync
wait 3 TransactionStatus.ACTIVE
wait 4 TransactionStatus.ACTIVE
wait 5 TransactionStatus.ACTIVE
B	4	ParseComplete
B	4	BindComplete
B	4	NoData
B	10	CommandComplete	 "BEGIN"
B	5	ReadyForQuery	 T
F	25	Query	 "select pg_sleep(0.1)"
cancelled
conn.info.transaction_status=<TransactionStatus.ACTIVE: 1>
F	4	Terminate
```

so it looks like the server ignores the cancel requests arriving precisely in that moment of the conversation. Not sure about that, but it might be something to investigate.

Other observations:

- the traces provided so far are on non-autocommit connections, so there is the BEGIN execution. Testing with autocommit is a different story.
- the query has no parameters: querying with parameters (`conn.execute("select pg_sleep(%s)", [0.1])
`) is also a different story.
- all the stories result in the connection remaining ACTIVE:

  - autocommit, simple protocol:

```
wait 1 TransactionStatus.IDLE
F	25	Query	 "select pg_sleep(0.1)"
wait 3 TransactionStatus.ACTIVE
wait 4 TransactionStatus.ACTIVE
wait 5 TransactionStatus.ACTIVE
B	33	RowDescription	 1 "pg_sleep" NNNN 0 NNNN 4 -1 0
B	NN	ErrorResponse	 S "ERROR" V "ERROR" C "57014" M "canceling statement due to user request" F "SSSS" L "SSSS" R "SSSS" \x00
cancelled
conn.info.transaction_status=<TransactionStatus.ACTIVE: 1>
F	4	Terminate
```

  - autocommit, extended protocol:

```
wait 1 TransactionStatus.IDLE
F	31	Parse	 "" "select pg_sleep($1)" 1 NNNN
F	28	Bind	 "" "" 1 1 1 8 '?\xffffffb9\xffffff99\xffffff99\xffffff99\xffffff99\xffffff99\xffffff9a' 1 0
F	6	Describe	 P ""
F	9	Execute	 "" 0
F	4	Sync
wait 3 TransactionStatus.ACTIVE
wait 4 TransactionStatus.ACTIVE
wait 5 TransactionStatus.ACTIVE
B	4	ParseComplete
B	4	BindComplete
B	33	RowDescription	 1 "pg_sleep" NNNN 0 NNNN 4 -1 0
B	NN	ErrorResponse	 S "ERROR" V "ERROR" C "57014" M "canceling statement due to user request" F "SSSS" L "SSSS" R "SSSS" \x00
cancelled
conn.info.transaction_status=<TransactionStatus.ACTIVE: 1>
F	4	Terminate
```

It seems, indeed, that the server receives the cancel instruction, and sends the message back to the client. However the status remains ACTIVE (it should be INERROR after receiving the error response).

@dlax do you have in mind cases in which the client receives an ErrorResponse but still it remains in ACTIVE transaction state?

**Comment by mikec1z** [comment]
I realize my example is flawed now regardless. I didn't see that `wait_async` was being called twice, so my example issues two cancellations of the task, the latter of which is during the handling of the first cancellation.  My use case doesn't issue multiple task cancellations this isn't reflective of my actual usage of psycopg. 

It would be suitable for my use case to have a means of cancelling the transaction after the interruption. I can check if the TransactionStatus is active and issue a cancellation. Unfortunately `connection.cancel` seems to have no effect. A simple option is to just close and reopen the connection but I prefer to avoid the overhead of this if possible.

**Comment by mikec1z** [comment]
I believe I have a suitable workaround for my use case.  If I call `connection.cancel` before cancelling the task, it will cause any queries to immediately exit. Close does nothing after `curr.execute` has returned (even if it left a query in process due to `asyncio.cancel`). Since cancel is blocking we can put it in a thread executor per your suggestion in  [#543](https://github.com/psycopg/psycopg/issues/534). 

In the event that no query is being run at the time of cancellation, we skip `connection.cancel` still but cancel the task so it will raise `CancelledError` on the next await.  

ex,

```
def blocking_cancel(conn):
    conn.cancel()

async def handle_request(cursor):
    try:
        await cursor.execute("select pg_sleep(100)")
    except psycopg.errors.QueryCanceled as e:
        pass # expected

async def main(...):
    ....
    with concurrent.futures.ThreadPoolExecutor() as pool:
        task = asyncio.create_task(handle_request(cursor))
        loop = asyncio.get_running_loop()
        await asyncio.sleep(.1) # give task a time to start for this example
        # if a query is being run cancel it
        if connection.info.transaction_status == TransactionStatus.ACTIVE:
            await loop.run_in_executor(pool, blocking_cancel, connection)
        task.cancel()
        try:
            await task
        except asyncio.CancelledError as e:
            pass #expected

        await connection.rollback()
        print(connection.info.transaction_status)

```

Output

> TransactionStatus.IDLE :)

**Reply by dvarrazzo** [reply]
>  If I call `connection.close` before cancelling the task

did you mean "if I call `connection.cancel`?

**Reply by mikec1z** [reply]
Yes, fixed. Although I think I've come up with something simpler involving `asyncio.shield` and not relying on `connection.cancel`

**Reply by dvarrazzo** [reply]
If there is any critical section for which we might use shield in psycopg code please let us know.

**Reply by mikec1z** [reply]
Ok I can confirm all the problems after the 3.1.9 patch were on my side. We have a web app where we need to cancel request handler tasks if the user disconnects prematurely. The network side of the code was sending multiple cancellations when it detected a disconnect, which was causing all the problems I saw after the patch. Thanks for your help.
